app/src/main/java/org/thoughtcrime/securesms/onboarding/LinkDeviceActivity.kt

```diff
@@ -1,230 +1,160 @@
-import org.thoughtcrime.securesms.util.setUpActionBarSessionLogo
+import org.thoughtcrime.securesms.ui.AppTheme
+import org.thoughtcrime.securesms.ui.baseBold
+import org.thoughtcrime.securesms.ui.components.MaybeScanQrCode
+import org.thoughtcrime.securesms.ui.components.OutlineButton
+import org.thoughtcrime.securesms.ui.components.SessionOutlinedTextField
+import org.thoughtcrime.securesms.ui.components.SessionTabRow
+import org.thoughtcrime.securesms.ui.contentDescription
 import javax.inject.Inject
 
+private const val TAG = "LinkDeviceActivity"
+
+private val TITLES = listOf(R.string.sessionRecoveryPassword, R.string.qrScan)
+
 @AndroidEntryPoint
-class LinkDeviceActivity : BaseActionBarActivity(), ScanQRCodeWrapperFragmentDelegate {
+@androidx.annotation.OptIn(ExperimentalGetImage::class)
+class LinkDeviceActivity : BaseActionBarActivity() {
 
     @Inject
-    lateinit var configFactory: ConfigFactory
+    lateinit var prefs: TextSecurePreferences
 
-    private lateinit var binding: ActivityLinkDeviceBinding
-    internal val database: LokiAPIDatabaseProtocol
-        get() = SnodeModule.shared.storage
-    private val adapter = LinkDeviceActivityAdapter(this)
-    private var restoreJob: Job? = null
+    val viewModel: LinkDeviceViewModel by viewModels()
 
-    @Deprecated("Deprecated in Java")
-    override fun onBackPressed() {
-        if (restoreJob?.isActive == true) return // Don't allow going back with a pending job
-        super.onBackPressed()
-    }
-
-    // region Lifecycle
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
-        setUpActionBarSessionLogo()
-        TextSecurePreferences.apply {
-            setHasViewedSeed(this@LinkDeviceActivity, true)
-            setConfigurationMessageSynced(this@LinkDeviceActivity, false)
-            setRestorationTime(this@LinkDeviceActivity, System.currentTimeMillis())
-            setLastProfileUpdateTime(this@LinkDeviceActivity, 0)
-        }
-        binding = ActivityLinkDeviceBinding.inflate(layoutInflater)
-        setContentView(binding.root)
-        binding.viewPager.adapter = adapter
-        binding.tabLayout.setupWithViewPager(binding.viewPager)
-    }
-    // endregion
-
-    // region Interaction
-    override fun handleQRCodeScanned(mnemonic: String) {
-        try {
-            val seed = Hex.fromStringCondensed(mnemonic)
-            continueWithSeed(seed)
-        } catch (e: Exception) {
-            Log.e("Loki","Error getting seed from QR code", e)
-            Toast.makeText(this, "An error occurred.", Toast.LENGTH_LONG).show()
+        supportActionBar?.setTitle(R.string.activity_link_load_account)
+        prefs.setHasViewedSeed(true)
+        prefs.setConfigurationMessageSynced(false)
+        prefs.setRestorationTime(System.currentTimeMillis())
+        prefs.setLastProfileUpdateTime(0)
+
+        lifecycleScope.launch {
+            viewModel.eventFlow.collect {
```

Comment:
Shouldn't you filter by event type? I know there is currently only one here, but you seem to have applied the filter in other parts even for single events. Good practice if it gets modified in the future.

Reply:
@bemusementpark

Reply:
not sure if that's the case. Filtering on if something is the type its defined as would probably give a warning for always true 🤔 so I think we leave it.